tests/cpp/test_multidevice_overlap.cpp

```diff
@@ -495,4 +495,134 @@ TEST_F(
+        params.M / params.S, params.N};
+
+    // Set up input tensors. We create the full unsharded tensors and define the
+    // actual input as the shard corresponding to the current device. Having the
+    // full unsharded input on each rank makes it possible to compute the
+    // expected result locally, hence, this way of doing is convenient for
+    // validating data correctness.
+    auto cpu_options = at::TensorOptions().dtype(at::kFloat);
+    at::TensorOptions gpu_options = cpu_options.device(communicator_->device());
+
+    ta_ = at::empty(ta_sizes, gpu_options);
+    tb_ = at::empty(tb_sizes, gpu_options);
+    tc_ = at::empty(tc_sizes, gpu_options);
+    ta_sharded_ = at::empty(ta_sharded_sizes, gpu_options);
+    tc_sharded_ = at::empty(tc_sharded_sizes, gpu_options);
+
+    // Debug print
+    if (communicator_->deviceId() == 0 && debug_print) {
+      debug() << "ta_sizes()=" << ta_.sizes() << std::endl
+              << "tb_sizes()=" << tb_.sizes() << std::endl
+              << "tc_sizes()=" << tc_.sizes() << std::endl;
+    }
+  }
+
+  // Each rank calls uniform_ and gets the same values for ta_ and tb_ because the random seed is initialized the same
+  // Therefore, we do not need to have one rank generate ta_ and tb_ and broadcast it to the rest of the ranks
+  void initializeIO() {
+    ta_.uniform_();
+    tb_.uniform_();
+    ta_sharded_.copy_(ta_.select(1, my_device_index_));
+  }
+
+  void validate() {
+    // compute the expected output for data correctness validation
+    auto tb_cpu = tb_.to(torch::kCPU);
+    auto ta_cpu = ta_.to(torch::kCPU);
+    auto tc_expected_ = torch::matmul(ta_cpu, tb_cpu);
+    EXPECT_TRUE(tc_.to(torch::kCPU).allclose(tc_expected_, 1e-1, 1e-1))
+        << "Unexpected results, obtained:" << tc_
+        << "\n expected: " << tc_expected_;
+  }
+};
+// This test implements an allgather-based pipelining overlapping technique,
+// similar to the above reduce-scattered based pipelining overlapping technique
+TEST_F(AGOverlapTest, AllgatherBasedPipeliningATenImplementation) {
+  std::vector<c10::cuda::CUDAStream> streams;
+  std::generate_n(
+      std::back_inserter(streams),
+      params.number_of_streams,
+      [my_device_index = my_device_index_]() {
+        return c10::cuda::getStreamFromPool(
+            /*isHighPriority=*/false, my_device_index);
+      });
+
+  for ([[maybe_unused]] const auto& _ :
+       c10::irange(params.number_of_iterations)) {
+    initializeIO();
+
+    for (auto j : c10::irange(params.S)) {
+      int64_t stream_index = j % streams.size();
+      setCurrentCUDAStream(streams.at(stream_index));
+
+      // local compute
+      auto ta_sharded_j = ta_sharded_.select(0, j); // params.M / (num_devices_ * params.S), params.K
+      auto tc_sharded_j = tc_sharded_.select(0, j); // params.M / (num_devices_ * params.S), params.N
+      torch::matmul_out(tc_sharded_j, ta_sharded_j, tb_); // params.M / (num_devices_ * params.S), params.N
```

Comment:
Can we permute `matmul` and `allgather`? I mean, to first allgather and then matmul. I sorry if I advised otherwise before.

I understand both solutions are possible, but since what we target is Megatron parallelization, we should stick to it and first study the AG+GEMM. Sorry for the trouble